R/build_projection_model.r

```diff
@@ -0,0 +1,64 @@
+#' Model future uptake from current uptake
+#'
+#' BEWARE: THIS DOES NOT YET ACCOUNT FOR MULTI-YEAR OR MULTI-STATE DATA SETS
+#'
+#' Future incident uptake can be predicted from current incident uptake.
+#' A model can be built and fit to project incident uptake into the future.
+#'
+#' The default model says that current incident uptake is a linear function
+#' of the previous incident uptake, time elapsed since rollout, and their
+#' interaction.
+#'
+#' All modeling is performed on the daily-average incident uptake to
+#' remove accumulation dependence among data points and to
+#' account for slight variations in incident reporting intervals.
+#' The first data point may be dropped before fitting the model
+#' if there is a long lag between the vaccine rollout date and
+#' the date that data collection began.
+#' All variables are standardized prior to fitting the model.
+#'
+#' UPDATES FOR THE FUTURE:
+#' - Model must incorporate year-to-year variability
+#' - Model must incorporate state-to-state variability
+#' - A variety of models should be tried and competed, or formula provided
+#'
+#' @description Fit a model to project uptake into the future
+#'
+#' @param data uptake data to fit the projection model
+#' @param drop_first whether to drop the first data point
+#'
+#' @return a brms model object
+#'
+#' @export
+#'
+build_projection_model <- function(data, drop_first = NULL) {
+
+  if (is.null(drop_first)) {
+    intervals <- diff(data$elapsed)
+    if (abs(data$elapsed[1] - mean(intervals)) > (sd(intervals) + 1)) {
```

Comment:
In `get_uptake_data.R`, the user defines a `start_date` which is the rollout date. `elapsed` is the number of days since that rollout date. Generally the first data point is after the rollout date, and sometimes by a lot. Thus, the first `elapsed` represents a time interval perhaps longer than most other intervals (which are often 7 days, but sometimes 6 or 8). I want to make sure the first interval is not much more than 1 standard deviation bigger than the others. This condition naturally becomes less strict if there is a lot of variation in the intervals (which could happen in other data sets, though not the data sets we are working with currently). But the opposite concern is if all the intervals are exactly 7 days, for example. Then, sd = 0, so if the first interval is 6 or 8 days, you still don't use it, which may not be reasonable. The "sd + 1" builds in a little tolerance for this.